monai/data/image_writer.py

```diff
@@ -0,0 +1,395 @@
+            align_corners: boolean option of ``grid_sample`` to handle the corner convention.
+                See also: https://pytorch.org/docs/stable/nn.functional.html#grid-sample
+            dtype: data type for resampling computation. Defaults to
+                ``np.float64`` for best precision. If ``None``, use the data type of input data.
+        """
+        data: np.ndarray
+        data, *_ = convert_data_type(data_array, np.ndarray)  # type: ignore
+
+        sr = min(data.ndim, 3)
+        if affine is None:
+            affine = np.eye(4, dtype=np.float64)
+        affine = to_affine_nd(sr, affine)  # type: ignore
+        target_affine = to_affine_nd(sr, target_affine) if target_affine is not None else affine
+
+        if np.allclose(affine, target_affine, atol=AFFINE_TOL):
+            # no affine changes, return (data, affine)
+            return data, ensure_mat44(target_affine)
+
+        # resolve orientation
+        if has_nib:  # this is to avoid dependency on nibabel
+            data, _affine = reorient_spatial_axes(data, affine, target_affine)
+            if np.allclose(_affine, target_affine, atol=AFFINE_TOL):
+                return data, ensure_mat44(_affine)
+
+        # need resampling
+        dtype = dtype or data.dtype  # type: ignore
+        if output_spatial_shape is None:
+            output_spatial_shape, _ = compute_shape_offset(data.shape, _affine, target_affine)
+        output_spatial_shape_ = list(output_spatial_shape) if output_spatial_shape is not None else []
+        sp_dims = min(data.ndim, 3)
+        output_spatial_shape_ += [1] * (sp_dims - len(output_spatial_shape_))
+        output_spatial_shape_ = output_spatial_shape_[:sp_dims]
+        original_channels = data.shape[3:]
+        if original_channels:  # multi channel, resampling each channel
+            data_np: np.ndarray = data.reshape(list(data.shape[:3]) + [-1])  # type: ignore
+            data_np = np.moveaxis(data_np, -1, 0)  # channel first for pytorch
+        else:  # single channel image, need to expand to have a channel
+            data_np = data[None]
+        affine_xform = AffineTransform(
+            normalized=False, mode=mode, padding_mode=padding_mode, align_corners=align_corners, reverse_indexing=True
+        )
+        data_torch = affine_xform(
+            torch.as_tensor(np.ascontiguousarray(data_np, dtype=dtype)).unsqueeze(0),
+            torch.as_tensor(np.ascontiguousarray(np.linalg.inv(_affine) @ target_affine, dtype=dtype)),
+            spatial_size=output_spatial_shape_,
+        )
+        data_np = data_torch[0].detach().cpu().numpy()
+        if original_channels:
+            data_np = np.moveaxis(data_np, 0, -1)  # channel last
+            data_np = data_np.reshape(list(data_np.shape[:3]) + list(original_channels))
+        else:
+            data_np = data_np[0]
+        return data_np, ensure_mat44(target_affine)
+
+    @classmethod
+    def convert_to_channel_last(
```

Comment:
I feel it's better to change this logic to the transform `AsChannelLast`, then users can apply below 3 transforms in postprocessing:
```py
AffineResample()  # the resample logic in this PR
AsChannelLast()
SaveImage()
```
What do you think?
@ericspod @rijobro for more discussions.

Thanks.

Reply:
The current design is the other way around: `SaveImage` as a transform should work well with the channel-first format, so that we hide the backend implementation details. For example, we can `SaveImage` to 'dump' intermediate transform outcomes conveniently. `SaveImage` internally calls resample/channel_last steps based on the requirements of the writer backend, not all backend consume the same channel last format.

Reply:
Hi @wyli ,

Thanks for your explanation, I see your design point.
But I am not very sure which way is better to go, as one of our MONAI design principles is "easy to integrate", maybe we should keep every component simple and independent? Daguang also raised the idea to extract `resample` from `SaveImage` during Clara 4.0 development.
@ericspod @rijobro Could you guys please help share your opinions?

Thanks in advance.

Reply:
I'm trying to create a `SpatialResample` transform, and `SaveImage` will internally call this transform if the user runs `SaveImage(resample=True)`， the user would also have the `SpatialResample` transform. I don't see why `SaveImage` shouldn't integrate the resampling functionality.

Reply:
Hi @wyli ,

That makes sense to me, if also providing `SpatialResample` transform, it's OK to call it in `SaveImage`.
And so this `resample` logic would be in the `SpatialResample` transform instead of this `ImageWriter` module, right?

Thanks.